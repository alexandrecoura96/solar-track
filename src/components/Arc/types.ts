export interface ArcProps {
  content?: string | null;
}
