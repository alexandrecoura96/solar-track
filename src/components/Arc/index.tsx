import numeral from 'numeral';
import React from 'react';
import {Dimensions} from 'react-native';
import {RFValue} from 'react-native-responsive-fontsize';
import Svg, {Defs, LinearGradient, Path, Stop, Text} from 'react-native-svg';
import {ArcProps} from './types';

const {width} = Dimensions.get('window');
const size = width - 72;
const strokeWidth = 16;
const strokeWidth2 = 32;
const {PI, cos, sin} = Math;
const r = (size - strokeWidth) / 2;
const cx = size / 2;
const cy = size / 2;
const A = PI + PI * 0.4;
const startAngle = PI + PI * 0.2;
const endAngle = 2 * PI - PI * 0.2;
const x1 = cx - r * cos(startAngle);
const y1 = -r * sin(startAngle) + cy;
const x2 = cx - r * cos(endAngle);
const y2 = -r * sin(endAngle) + cy;
const d = `M ${x1} ${y1} A ${r} ${r} 0 1 0 ${x2} ${y2}`;

const startAngle2 = PI + PI * 0.2;
const endAngle2 = 2 * PI - PI * 0.2;
const x1_2 = cx - (r - strokeWidth2 / 1) * cos(startAngle2);
const y1_2 = -(r - strokeWidth2 / 1) * sin(startAngle2) + cy;
const x2_2 = cx - (r - strokeWidth2 / 1) * cos(endAngle2);
const y2_2 = -(r - strokeWidth2 / 1) * sin(endAngle2) + cy;
const d2 = `M ${x1_2} ${y1_2} A ${r - strokeWidth2 / 1} ${
  r - strokeWidth2 / 1
} 0 1 0 ${x2_2} ${y2_2}`;

export const Arc = ({content}: ArcProps) => {
  const circumference = r * A;
  const α = 100 * A;

  const strokeDashoffset = α * r;

  return (
    <Svg width={size} height={RFValue(230)}>
      <Defs>
        <LinearGradient id="grad" x1="0" y1="0" x2="100%" y2="0">
          <Stop offset="0" stopColor="#fbf0ce" />
          <Stop offset="1" stopColor="#dfaa6d" />
        </LinearGradient>
      </Defs>

      <Path
        stroke="#ef9837"
        strokeLinecap="round"
        fill="none"
        strokeDasharray={`${circumference}, ${circumference}`}
        {...{d, strokeWidth}}
      />

      <Path
        fill="none"
        strokeDasharray={`${circumference}, ${circumference}`}
        strokeDashoffset={strokeDashoffset}
        strokeLinecap="round"
        {...{d, strokeWidth}}
      />
      {content && (
        <Text
          fill="#fff"
          fontSize={32}
          fontFamily="Inter-Bold"
          textAnchor="middle"
          lengthAdjust="spacingAndGlyphs"
          x={size / 2}
          y={size / 2 + 10}>
          {numeral(content).format().slice(0, 10)}
        </Text>
      )}

      <Path
        stroke="url(#grad)"
        fill="none"
        opacity={0.8}
        strokeLinecap="round"
        strokeWidth={28}
        strokeDasharray={`${circumference}, ${circumference}`}
        {...{d: d2}}
      />
    </Svg>
  );
};
