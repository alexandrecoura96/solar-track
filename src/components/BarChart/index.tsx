import {
  Canvas,
  Path,
  Skia,
  Text,
  runTiming,
  useComputedValue,
  useFont,
  useValue,
} from '@shopify/react-native-skia';
import * as d3 from 'd3';
import numeral from 'numeral';
import React from 'react';
import {Button, Easing, StyleSheet, View} from 'react-native';
import {useSharedValue, withTiming} from 'react-native-reanimated';
import {BarChartDataProps, BarChartProps} from './types';

const GRAPH_MARGIN = 20;
const GRAPH_BAR_WIDTH = 8;

const CanvasHeight = 350;
const CanvasWidth = 350;
const graphHeight = CanvasHeight - 2 * GRAPH_MARGIN;
const graphWidth = CanvasWidth - 2;

export const BarChart = ({data}: BarChartProps) => {
  const font = useFont(require('../../assets/fonts/Inter-Regular.ttf'), 10);
  const animationState = useValue(0);
  const opacityAnimValue = useSharedValue(0);

  const xDomain = data.map((dataPoint: BarChartDataProps) => dataPoint.label);
  const xRange = [0, graphWidth];
  const x = d3.scalePoint().domain(xDomain).range(xRange).padding(1);

  const yDomain: number[] = [
    0,
    d3.max(
      data,
      (yBarChartDataProps: BarChartDataProps) => yBarChartDataProps.value,
    )!,
  ];

  const yRange = [0, graphHeight];
  const y = d3.scaleLinear().domain(yDomain).range(yRange);

  const animate = () => {
    animationState.current = 0;

    runTiming(animationState, 1, {
      duration: 1600,
      easing: Easing.inOut(Easing.exp),
    });

    opacityAnimValue.value = withTiming(1, {
      duration: 1600,
    });
  };

  const path = useComputedValue(() => {
    const newPath = Skia.Path.Make();

    data.forEach((dataPoint: BarChartDataProps) => {
      const rect = Skia.XYWHRect(
        x(dataPoint.label)! - GRAPH_BAR_WIDTH / 2,
        graphHeight,
        GRAPH_BAR_WIDTH,
        y(dataPoint.value * animationState.current) * -1,
      );

      const rrect = Skia.RRectXY(rect, 8, 8);
      newPath.addRRect(rrect);
    });

    return newPath;
  }, [animationState]);

  if (!font) {
    return <View />;
  }

  return (
    <View style={styles.container}>
      <Canvas style={styles.canvasTitle}>
        {data.map((dataPoint: BarChartDataProps) => (
          <Text
            key={dataPoint.label}
            color="#fff"
            font={font}
            x={x(dataPoint.label)! - 10}
            y={CanvasHeight - 60}
            text={numeral(dataPoint.value).format('0a')}
            opacity={opacityAnimValue}
          />
        ))}
      </Canvas>

      <Canvas style={styles.canvas}>
        <Path path={path} color="#e05f25" />

        {data.map((dataPoint: BarChartDataProps) => (
          <Text
            color="#fff"
            key={dataPoint.label}
            font={font}
            x={x(dataPoint.label)! - 10}
            y={CanvasHeight - 25}
            text={dataPoint.label}
          />
        ))}
      </Canvas>
      <Button title="Animate!" onPress={animate} />
    </View>
  );
};

const styles = StyleSheet.create({
  container: {
    justifyContent: 'center',
    alignItems: 'center',
    flex: 0.5,
    backgroundColor: '#000',
  },
  canvasTitle: {
    height: CanvasHeight - 20,
    width: CanvasWidth,
    paddingBottom: 60,
  },
  canvas: {
    height: CanvasHeight,
    width: CanvasWidth,
  },
});
