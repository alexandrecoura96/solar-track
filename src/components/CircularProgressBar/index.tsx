import {
  Easing,
  SkFont,
  runTiming,
  useFont,
  useValue,
} from '@shopify/react-native-skia';
import React, {useCallback, useEffect} from 'react';
import {PixelRatio} from 'react-native';
import {DonutChart} from './DonutChart';
import {CircularProgressBarProps} from './types';

export function CircularProgressBar({percentage}: CircularProgressBarProps) {
  const font = useFont(require('../../assets/fonts/Inter-Black.ttf'), 16);
  const radius = PixelRatio.roundToNearestPixel(36);
  const STROKE_WIDTH = 5;
  const animationState = useValue(0);
  const targetPercentage = percentage / 100;

  const animateChart = useCallback(() => {
    animationState.current = 0;
    runTiming(animationState, targetPercentage, {
      duration: 1250,
      easing: Easing.inOut(Easing.cubic),
    });
  }, [animationState, targetPercentage]);

  useEffect(() => {
    animateChart();
  }, [animateChart]);

  if (!font) {
    return null;
  }

  return (
    <DonutChart
      radius={radius}
      strokeWidth={STROKE_WIDTH}
      percentageComplete={animationState}
      targetPercentage={targetPercentage}
      font={font as SkFont}
    />
  );
}
