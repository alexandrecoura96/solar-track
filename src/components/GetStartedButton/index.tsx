import React from 'react';
import {RFValue} from 'react-native-responsive-fontsize';
import Play from '../../assets/svg/play.svg';
import {GetStarted, GetStartedIconWrapper, GetStartedText} from './styles';
import {GetStartedButtonProps} from './types';

export function GetStartedButton({title, ...rest}: GetStartedButtonProps) {
  return (
    <GetStarted {...rest} activeOpacity={0.8}>
      <GetStartedText>{title}</GetStartedText>
      <GetStartedIconWrapper>
        <Play height={RFValue(16)} width={RFValue(16)} />
      </GetStartedIconWrapper>
    </GetStarted>
  );
}
