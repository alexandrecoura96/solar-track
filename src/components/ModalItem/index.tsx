import React from 'react';
import {RFValue} from 'react-native-responsive-fontsize';
import Check from '../../assets/svg/check_circle.svg';
import {Container, Title} from './styles';
import {ModalItemProps} from './types';

export function ModalItem({
  title,
  selected,
  hasBorder = true,
  ...rest
}: ModalItemProps) {
  return (
    <Container hasBorder={hasBorder} activeOpacity={0.8} {...rest}>
      <Title>{title}</Title>
      {selected && (
        <Check fill="#1C1E1B" height={RFValue(20)} width={RFValue(20)} />
      )}
    </Container>
  );
}
