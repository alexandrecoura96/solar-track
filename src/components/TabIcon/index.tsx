import React from 'react';
import GraphicsIcon from '../../assets/svg/graphics.svg';
import HomeIcon from '../../assets/svg/home.svg';
import SettingsIcon from '../../assets/svg/settings.svg';

import {RFValue} from 'react-native-responsive-fontsize';
import {TabIconProps} from './types';

export const TabIcon = ({color}: TabIconProps) => {
  return {
    Home: <HomeIcon height={RFValue(24)} width={RFValue(24)} fill={color} />,
    Graphics: (
      <GraphicsIcon height={RFValue(24)} width={RFValue(24)} fill={color} />
    ),
    Settings: (
      <SettingsIcon height={RFValue(24)} width={RFValue(24)} fill={color} />
    ),
  };
};
