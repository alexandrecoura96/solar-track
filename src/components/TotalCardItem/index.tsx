import React from 'react';
import {Container, Description, Title} from './styles';
import {TotalCardItemProps} from './types';

export function TotalCardItem({title, description}: TotalCardItemProps) {
  return (
    <Container>
      <Description>{description ? description : '0'}</Description>
      {title && <Title>{title}</Title>}
    </Container>
  );
}
