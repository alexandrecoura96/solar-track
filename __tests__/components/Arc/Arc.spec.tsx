import React from 'react';
import renderer from 'react-test-renderer';
import {Arc} from '../../../src/components/Arc';

describe('Arc', () => {
  test('Should render tree of snapshot test in component', () => {
    const tree = renderer.create(<Arc content="content" />);

    expect(tree).toMatchSnapshot();
  });
});
