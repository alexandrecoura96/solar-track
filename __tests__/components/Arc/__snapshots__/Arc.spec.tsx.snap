// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Arc Should render tree of snapshot test in component 1`] = `
<RNSVGSvgView
  bbHeight="451"
  bbWidth="678"
  focusable={false}
  height={451}
  style={
    [
      {
        "backgroundColor": "transparent",
        "borderWidth": 0,
      },
      {
        "flex": 0,
        "height": 451,
        "width": 678,
      },
    ]
  }
  width={678}
>
  <RNSVGGroup
    fill={
      {
        "payload": 4278190080,
        "type": 0,
      }
    }
  >
    <RNSVGDefs>
      <RNSVGLinearGradient
        gradient={
          [
            0,
            -266034,
            1,
            -2119059,
          ]
        }
        gradientTransform={null}
        gradientUnits={0}
        name="grad"
        x1="0"
        x2="100%"
        y1="0"
        y2="0"
      />
    </RNSVGDefs>
    <RNSVGPath
      d="M 606.7846251381077 533.5569185088086 A 331 331 0 1 0 71.21537486189243 533.5569185088086"
      fill={null}
      propList={
        [
          "fill",
          "stroke",
          "strokeWidth",
          "strokeDasharray",
          "strokeLinecap",
        ]
      }
      stroke={
        {
          "payload": 4293892151,
          "type": 0,
        }
      }
      strokeDasharray={
        [
          "1455.81403567351",
          "1455.81403567351",
        ]
      }
      strokeLinecap={1}
      strokeWidth="16"
    />
    <RNSVGPath
      d="M 606.7846251381077 533.5569185088086 A 331 331 0 1 0 71.21537486189243 533.5569185088086"
      fill={null}
      propList={
        [
          "fill",
          "strokeWidth",
          "strokeDasharray",
          "strokeDashoffset",
          "strokeLinecap",
        ]
      }
      strokeDasharray={
        [
          "1455.81403567351",
          "1455.81403567351",
        ]
      }
      strokeDashoffset={145581.40356735102}
      strokeLinecap={1}
      strokeWidth="16"
    />
    <RNSVGText
      content={null}
      dx={[]}
      dy={[]}
      fill={
        {
          "payload": 4294967295,
          "type": 0,
        }
      }
      font={
        {
          "fontFamily": "Inter-Bold",
          "fontSize": "32",
          "textAnchor": "middle",
        }
      }
      propList={
        [
          "fill",
        ]
      }
      rotate={[]}
      x={
        [
          "339",
        ]
      }
      y={
        [
          "349",
        ]
      }
    >
      <RNSVGTSpan
        content="0"
        dx={[]}
        dy={[]}
        fill={
          {
            "payload": 4278190080,
            "type": 0,
          }
        }
        font={{}}
        rotate={[]}
        x={[]}
        y={[]}
      />
    </RNSVGText>
    <RNSVGPath
      d="M 580.8960813181093 514.7477904354495 A 299 299 0 1 0 97.10391868189075 514.7477904354496"
      fill={null}
      opacity={0.8}
      propList={
        [
          "fill",
          "stroke",
          "strokeWidth",
          "strokeDasharray",
          "strokeLinecap",
        ]
      }
      stroke={
        {
          "brushRef": "grad",
          "type": 1,
        }
      }
      strokeDasharray={
        [
          "1455.81403567351",
          "1455.81403567351",
        ]
      }
      strokeLinecap={1}
      strokeWidth="28"
    />
  </RNSVGGroup>
</RNSVGSvgView>
`;
